tests/DecisionServiceTests/DecisionServiceTest.php

```diff
@@ -603,6 +614,445 @@ public function testGetVariationWithBucketingId()
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeature($feature_flag, 'user_1', []),
+            null
+        );
+    }
+
+    public function testGetVariationForFeatureRolloutWhenNoRolloutIsAssociatedToFeatureFlag(){
+         $feature_flag = $this->config->getFeatureFlagFromKey('boolean_feature');
+
+         $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::DEBUG, 
+            "Feature flag 'boolean_feature' is not used in a rollout.");
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeatureRollout($feature_flag, 'user_1', []),
+            null
+        );
+    }
 
+
+    public function testGetVariationForFeatureRolloutWhenRolloutIsNotInDataFile(){
+         $feature_flag = $this->config->getFeatureFlagFromKey('boolean_feature');
+         $feature_flag = clone $feature_flag;
+         $feature_flag->setRolloutId('invalid_rollout_id');
+
+         $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::ERROR, 
+            'Rollout with ID "invalid_rollout_id" is not in the datafile.');
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeatureRollout($feature_flag, 'user_1', []),
+            null
+        );
+    }
+
+    public function testGetVariationForFeatureRolloutWhenRolloutDoesNotHaveExperiment(){
+        // Mock Project Config
+        $configMock = $this->getMockBuilder(ProjectConfig::class)
+            ->setConstructorArgs(array(DATAFILE, $this->loggerMock, new NoOpErrorHandler()))
+            ->setMethods(array('getRolloutFromId'))
+            ->getMock(); 
+
+        $this->decisionService = new DecisionService($this->loggerMock, $configMock);
+  
+        $feature_flag = $this->config->getFeatureFlagFromKey('boolean_single_variable_feature');
+        $rollout_id = $feature_flag->getRolloutId();
+        $rollout = $this->config->getRolloutFromId($rollout_id);
+        $experiment_less_rollout = clone $rollout;
+        $experiment_less_rollout->setExperiments([]);
+
+        $configMock
+        ->method('getRolloutFromId')
+        ->will($this->returnValue($experiment_less_rollout));
+
+        $this->assertEquals(
+            $this->decisionService->getVariationForFeatureRollout($feature_flag, 'user_1', []),
+            null
+        );
+    }
+
+    // ============== when the user qualifies for targeting rule (audience match) ======================
+    
+    //  should return the variation the user is bucketed into when the user is bucketed into the targeting rule
+    public function testGetVariationForFeatureRolloutWhenUserIsBucketedInTheTargetingRule(){
+        $feature_flag = $this->config->getFeatureFlagFromKey('boolean_single_variable_feature');
+        $rollout_id = $feature_flag->getRolloutId();
+        $rollout = $this->config->getRolloutFromId($rollout_id);
+        $experiment = $rollout->getExperiments()[0];
```

Comment:
please change it to roleoutRule, it's really confusing me.

Reply:
In dynamic languages i think this should be experiment. So that it's clear to developers, the type of object it has,